core/cache/resident.go

```diff
@@ -79,8 +86,43 @@ func (m *residentManager) new() *Resident {
 	return r
 }
 
-func (m *residentManager) evict(id uint64) {
-	// TODO (manadart 2019-04-17): TBC when the mark/sweep logic is added.
+// mark sets all of the manager's residents to be stale.
+func (m *residentManager) mark() {
+	for _, r := range m.residents {
+		r.stale = true
+	}
+}
+
+// sweep removes stale cache residents in descending order of ID.
+// Because the IDs are supplied in increasing order, this ensures
+// we never remove a resident's model before the resident itself.
+func (m *residentManager) sweep() error {
+	// Create a descending order slice of IDs.
+	residentIds := make([]uint64, len(m.residents))
+	i := 0
+	for id := range m.residents {
+		residentIds[i] = id
+		i++
+	}
+	sort.Sort(uint64Reverse(residentIds))
+
+	// Read the map "leaves-first" and evict stale residents.
+	for _, id := range residentIds {
+		r := m.residents[id]
+		if r.stale {
+			if r.removalMessage == nil {
+				logger.Warningf("cache resident %d has no removal message; skipping eviction", id)
+				continue
+			}
+
+			select {
+			case m.removals <- r.removalMessage:
+			case <-time.After(10 * time.Second):
```

Comment:
Somewhat arbitrary; welcome suggestions.

Reply:
We should initialize the residentManager with a "done" channel, that is the controller tomb Dying() channel. That way we don't need a time here, we could just use the "done" channel.